Document Request.signal (#20843)
cloudflare/cloudflare-docs: src/content/changelog/workers/2025-03-14-handle-request-cancellation.mdx

---

> title: Handle request cancellation with Request.signal

Reviewer: ```suggestion
title: Handle incoming request cancellation in Workers with Request.signal
```

Workers just because it's a global Cloudflare-wide changelog, and "incoming" because my understanding is that's the specific change we've made here?

description: Workers can now add event listeners on Request.signal and perform tasks when the request is cancelled
products:
  - workers

> date: 2025-03-14T01:00:00Z

Author: It's still deploying now - should we set this date to Monday?

Reviewer: Yep not a rush

---

import { Render, TypeScriptExample } from "~/components";

> `Request` objects now have a `signal` property. You can now attach an event listener to the `signal` and perform tasks when the request to your Worker is canceled by the client.

Reviewer: ```suggestion
In Cloudflare Workers, you can now attach an event listener to [`Request`](/workers/runtime-apis/request/) objects, using the [`signal` property](https://developer.mozilla.org/en-US/docs/Web/API/Request/signal), and perform tasks when the request to your Worker is canceled by the client.
```

> This lets you do things like clean up after the request or write to logs before your Worker is aborted.

Reviewer: ```suggestion
This allows you to perform cleanup tasks or write to logs before your Worker's invocation ends. For example, if you run the Worker below, and then abort the request from the client, a log will be written:
```

<TypeScriptExample filename="index.ts">
```ts

> import { WorkerEntrypoint } from 'cloudflare:workers'; 

Reviewer: Might be simplest for example to use normal worker default export object with a fetch() handler, rather than introducing entrypoints? Just to keep reader focused on the thing that is new here

class Example extends WorkerEntrypoint {
  async fetch(req: Request): Response {
    req.signal.onabort = () => {
      console.log('The request was aborted!');
    };

    const { readable, writable } = new IdentityTransformStream();
    this.ctx.waitUntil(this.sendPing(writable));
    return new Response(readable);
  }

>   async sendPing(writable: WritableStream): void {

Author: My example is a bit contrived...

Reviewer: I like that it's runnable! Trying it now

Reviewer: I tried here:

https://github.com/irvinebroque/request-signal

in local dev when client aborts the request I don't see logs

Author: Thanks for verifying the example! I'll clone it locally and see if there's something wrong with it.

Author: A couple of different things going on here:
- wrangler hasn't yet been updated to a version of workerd that includes this change.
- Testing with a copy I updated locally, I found there were a couple of small mistakes in the example.

I'll push a fix.
